### PR TITLE
Notebooks respect `files.xyz` settings

### LINKED ISSUE #181905: Bug: Jupyter Notebook trim whitespaces not working in code cells
Does this issue occur when all extensions are disabled?: Yes
Note: All extensions disabled - except python and jupyter notebook 

The function to trim whitespaces does not work in jupyter notebook code cells

- VS Code Version:  1.78.0 (252e5463d60e63238250799aef7375787f68b4ee) - x64
- OS Version: Win 11 Pro

Steps to Reproduce:

1. Create a new Jupyter Notebook (.ipynb) file
2. Create a cell and put some code with tailing whitespaces and empty lines in it

Example:
image

### PR DESCRIPTION
**Pull request by Yoyokrazy**
Fixes: #181905

Notebooks now respect the settings:
- `files.trimTrailingWhitespace`
- `files.trimFinalNewlines`

### CODE REVIEW

**Comment by Yoyokrazy** [comment]
Also support:
- `files.insertFinalNewLine`

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 1]
```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+		// autosave -- don't trim entire line, only up to cursor, so need to track position of cursor(s)
+		if (context.reason === SaveReason.AUTO) {
+			viewCell = getNotebookViewCell(this.editorService);
+			if (!viewCell) {
+				return;
+			}
+			const selections = viewCell.getSelections();
+			for (const sel of selections) {
+				if (viewCell.model.textModel) {
+					cursors.push(new Position(sel.selectionStartLineNumber, sel.startColumn));
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:138]: If we are tracking the cursor it's `positionLineNumber` and `positionColumn`, not start*

Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:138]: We may want to follow https://github.com/microsoft/vscode/blob/7151639826490c7af63befbee10a0c229c47b020/src/vs/workbench/contrib/codeEditor/browser/saveParticipants.ts#L63-L71

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 2]
```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+			this.doTrimFinalNewLines(workingCopy, context, progress);
+		}
+	}
+
+	/**
+	 * returns 0 if the entire file is empty
+	 */
+	private findLastNonEmptyLine(model: ITextModel): number {
+		for (let lineNumber = model.getLineCount(); lineNumber >= 1; lineNumber--) {
+			const lineContent = model.getLineContent(lineNumber);
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:191]: `getLineContent` is too costly, we should check `lineLength`

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 3]
```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+		}
+
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
+					return;
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:233]: We don't need `model` here, `cell.textBuffer` is sufficient.

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 4]
```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+				disposable.add(ref);
+				const model = ref.object.textEditorModel;
+
+				const lineCount = model.getLineCount();
+				const lastLine = model.getLineContent(lineCount);
+				const lastLineIsEmptyOrWhitespace = strings.lastNonWhitespaceIndex(lastLine) === -1;
+
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:295]: Same here, let's use `textBuffer` so we don't have to create a temporary `textModel`. We can check if the last line is empty by comparing `getLineLastNonWhitespaceColumn` and `getLineMaxColumn`

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 5]
```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+		// autosave -- don't trim entire line, only up to cursor, so need to track position of cursor(s)
+		if (context.reason === SaveReason.AUTO) {
+			viewCell = getNotebookViewCell(this.editorService);
+			if (!viewCell) {
+				return;
+			}
+			const selections = viewCell.getSelections();
+			for (const sel of selections) {
+				if (viewCell.model.textModel) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:220]: Why do we need to check `viewCell.model.textModel` here?

Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:220]: My understanding is we want to cache editor `selection` as view states, and then use the `selection` view state from the active cell, it's fine if the cell is moved out of view.

**Review by rebornix — CHANGES_REQUESTED** [review]
It's in the right direction :thumbsup:, left some comments on how we can perform more efficiently.

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts (hunk 2 above)
Comment (chandra1109) [reply src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:191]: recheck with the process id 187787 to verify with the modules existing already

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts (hunk 3 above)
Comment (chandra1109) [reply src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:233]: Model need not be specified within the file available .
-> only cell.textbuffer would be sufficient for the function in order to run

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 6]
```diff
@@ -86,7 +93,207 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+			return;
+		}
+
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+		const activeCellEditor = getActiveCellCodeEditor(this.editorService);
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:214]: let's use `CellKind.Code`

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 7]
```diff
@@ -86,7 +93,207 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+				const isActiveCell = (activeCellEditor && cell.uri.toString() === activeCellEditor.getModel()?.uri.toString());
+				if (isAutoSaved && isActiveCell) {
+					const selections = activeCellEditor.getSelections() ?? [];
+					for (const sel of selections) {
+						cannotTouchLineNumber = Math.max(cannotTouchLineNumber, sel.selectionStartLineNumber);
+					}
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
+				disposable.add(ref);
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:229]: `ref` is no longer userd.

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 8]
```diff
@@ -86,7 +93,207 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
+					return;
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:282]: `ref` is no longer userd.

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 9]
```diff
@@ -86,7 +93,207 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
...
+		if (!workingCopy.model || !(workingCopy.model instanceof NotebookFileWorkingCopyModel)) {
+			return;
+		}
+
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:278]: Same here, let's use enum.

**Review by rebornix — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts (hunk 7 above)
Comment (chandra1109) [reply src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:229]: reference no longer needed to be used

**Review by rebornix — APPROVED** [review]
Good work, thank you!

**Comment by chandra1109** [comment]
#193524 
Review

File: src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts [hunk 10]
```diff
@@ -299,8 +503,11 @@ export class SaveParticipantsContribution extends Disposable implements IWorkben
 	}
 
 	private registerSaveParticipants(): void {
+		this._register(this.workingCopyFileService.addSaveParticipant(this.instantiationService.createInstance(TrimWhitespaceParticipant)));
 		this._register(this.workingCopyFileService.addSaveParticipant(this.instantiationService.createInstance(CodeActionOnSaveParticipant)));
 		this._register(this.workingCopyFileService.addSaveParticipant(this.instantiationService.createInstance(FormatOnSaveParticipant)));
+		this._register(this.workingCopyFileService.addSaveParticipant(this.instantiationService.createInstance(FinalNewLineParticipant)));
+		this._register(this.workingCopyFileService.addSaveParticipant(this.instantiationService.createInstance(TrimFinalNewLinesParticipant)));
 	}
 }
```
Comment (chandra1109) [inline src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts:512]: #193140  Top Padding is wrong in the given section of the header that is made available .

**Review by chandra1109 — COMMENTED** [review]
Review pending..

**Pull request merged**